server/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -1254,7 +1360,11 @@ private CanMatchResponse canMatch(ShardSearchRequest request, boolean checkRefre
             } else {
                 indexService = indicesService.indexServiceSafe(request.shardId().getIndex());
                 IndexShard indexShard = indexService.getShard(request.shardId().getId());
-                hasRefreshPending = indexShard.hasRefreshPending() && checkRefreshPending;
+                boolean needsWaitForRefresh = request.waitForCheckpoint() != SequenceNumbers.NO_OPS_PERFORMED;
+                // If this request requests wait_for_refresh behavior, it is safest to assume a refresh is pending. Theoretically,
+                // this can be improved in the future by manually checking that the requested checkpoint has already been refresh.
+                // However, this will request modifying the engine to surface that information.
+                hasRefreshPending = checkRefreshPending && (indexShard.hasRefreshPending() || needsWaitForRefresh);
```

Comment:
See comment above, either we need the refresh to happen earlier (inside `rewriteAndFetchShardRequest`) or we need to let `hasRefreshPending` be true when `needsWaitForRefresh=true` regarless of `checkRefreshPending`.

Reply:
I went with this approach.